From 9bddd543812fc68fc1cad6eb3129e48229418ce7 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Tue, 25 Jun 2024 18:25:18 +0200
Subject: [PATCH] Disable BCH/XEC upgrades (#13)

These introduce new limits on e.g. Script execution which Dogecoin doesn't have, and therefore need to be disabled.

Achieve this by setting the activation height to 0x7ffffffe (one test requires a +1).

The wellington and cowperthwaite upgrade keep a sensible height, which would otherwise break the tests, but we disable their effect by setting enableMinerFund and enableStakingRewards to false on mainnet. Therefore they are effectively disabled.

diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index a2920b70df0c86f8798dcdf84262ce71c5cd23e0..c80f47bacb2af889dcd9218a8a3279403b3c2140 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -111,11 +111,11 @@ public:
         // two days
         consensus.nDAAHalfLife = 2 * 24 * 60 * 60;
 
-        // The miner fund is enabled by default on mainnet.
-        consensus.enableMinerFund = true;
+        // The miner fund is disabled by default on Dogecoin mainnet.
+        consensus.enableMinerFund = false;
 
-        // The staking rewards are enabled by default on mainnet.
-        consensus.enableStakingRewards = true;
+        // The staking rewards are disabled by default on Dogecoin mainnet.
+        consensus.enableStakingRewards = false;
 
         // The best chain should have at least this much work.
         consensus.nMinimumChainWork =
@@ -126,35 +126,18 @@ public:
         consensus.defaultAssumeValid =
             ChainParamsConstants::MAINNET_DEFAULT_ASSUME_VALID;
 
-        // August 1, 2017 hard fork
-        consensus.uahfHeight = 478558;
-
-        // November 13, 2017 hard fork
-        consensus.daaHeight = 504031;
-
-        // November 15, 2018 hard fork
-        consensus.magneticAnomalyHeight = 556766;
-
-        // November 15, 2019 protocol upgrade
-        consensus.gravitonHeight = 609135;
-
-        // May 15, 2020 12:00:00 UTC protocol upgrade
-        consensus.phononHeight = 635258;
-
-        // Nov 15, 2020 12:00:00 UTC protocol upgrade
-        consensus.axionHeight = 661647;
-
-        // May 15, 2023 12:00:00 UTC protocol upgrade
-        consensus.wellingtonHeight = 792116;
-
-        // Nov 15, 2023 12:00:00 UTC protocol upgrade
-        consensus.cowperthwaiteHeight = 818669;
-
-        // Nov 15, 2026 12:00:00 UTC protocol upgrade
-        consensus.mengerActivationTime = 1794744000;
-
-        // May 15, 2027 12:00:00 UTC protocol upgrade
-        consensus.sandersActivationTime = 1810382400;
+        // BCH/XEC upgrades, disabled on Dogecoin
+        // Avoid using 0x7fffffff since some tests would overflow
+        consensus.uahfHeight = 0x7ffffffe;
+        consensus.daaHeight = 0x7ffffffe;
+        consensus.magneticAnomalyHeight = 0x7ffffffe;
+        consensus.gravitonHeight = 0x7ffffffe;
+        consensus.phononHeight = 0x7ffffffe;
+        consensus.axionHeight = 0x7ffffffe;
+        consensus.wellingtonHeight = 800000; // keep alive for tests
+        consensus.cowperthwaiteHeight = 900000; // keep alive for tests
+        consensus.mengerActivationTime = std::numeric_limits<int64_t>::max();
+        consensus.sandersActivationTime = std::numeric_limits<int64_t>::max();
 
         // Dogecoin: Digishield activation height
         consensus.digishieldHeight = 145000;
@@ -297,35 +280,17 @@ public:
         consensus.defaultAssumeValid =
             ChainParamsConstants::TESTNET_DEFAULT_ASSUME_VALID;
 
-        // August 1, 2017 hard fork
-        consensus.uahfHeight = 1155875;
-
-        // November 13, 2017 hard fork
-        consensus.daaHeight = 1188697;
-
-        // November 15, 2018 hard fork
-        consensus.magneticAnomalyHeight = 1267996;
-
-        // November 15, 2019 protocol upgrade
-        consensus.gravitonHeight = 1341711;
-
-        // May 15, 2020 12:00:00 UTC protocol upgrade
-        consensus.phononHeight = 1378460;
-
-        // Nov 15, 2020 12:00:00 UTC protocol upgrade
-        consensus.axionHeight = 1421481;
-
-        // May 15, 2023 12:00:00 UTC protocol upgrade
-        consensus.wellingtonHeight = 1556117;
-
-        // Nov 15, 2023 12:00:00 UTC protocol upgrade
-        consensus.cowperthwaiteHeight = 1584485;
-
-        // Nov 15, 2026 12:00:00 UTC protocol upgrade
-        consensus.mengerActivationTime = 1794744000;
-
-        // May 15, 2027 12:00:00 UTC protocol upgrade
-        consensus.sandersActivationTime = 1810382400;
+        // BCH/XEC upgrades, disabled on Dogecoin
+        consensus.uahfHeight = 0x7ffffffe;
+        consensus.daaHeight = 0x7ffffffe;
+        consensus.magneticAnomalyHeight = 0x7ffffffe;
+        consensus.gravitonHeight = 0x7ffffffe;
+        consensus.phononHeight = 0x7ffffffe;
+        consensus.axionHeight = 0x7ffffffe;
+        consensus.wellingtonHeight = 0x7ffffffe;
+        consensus.cowperthwaiteHeight = 0x7ffffffe;
+        consensus.mengerActivationTime = std::numeric_limits<int64_t>::max();
+        consensus.sandersActivationTime = std::numeric_limits<int64_t>::max();
 
         // Dogecoin: Digishield activation height
         consensus.digishieldHeight = 145000;
diff --git a/src/test/minerfund_tests.cpp b/src/test/minerfund_tests.cpp
index 94bbf49d2db88673b3869029aeab5666c5f60804..30d83dbc7e7b814d288ada1911bc5db637f803a4 100644
--- a/src/test/minerfund_tests.cpp
+++ b/src/test/minerfund_tests.cpp
@@ -31,6 +31,7 @@ CheckWhitelist(const Consensus::Params &consensusParams,
 BOOST_AUTO_TEST_CASE(minerfund_whitelist) {
     const CChainParams &chainparams = Params();
     const Consensus::Params &consensusParams = chainparams.GetConsensus();
+    Assert(m_node.args)->ForceSetArg("-enableminerfund", "1");
 
     std::array<CBlockIndex, 3> blocks;
 
@@ -53,6 +54,8 @@ BOOST_AUTO_TEST_CASE(minerfund_whitelist) {
 
     BOOST_CHECK(IsWellingtonEnabled(consensusParams, blocks[2].pprev));
     CheckWhitelist(consensusParams, &blocks[2], expectedMinerFund);
+
+    Assert(m_node.args)->ClearForcedArg("-enableminerfund");
 }
 
 BOOST_AUTO_TEST_CASE(minerfund_amount) {
diff --git a/src/test/policy_block_tests.cpp b/src/test/policy_block_tests.cpp
index 548f78695572250bc2f4e61787953fbebab21f14..0173ec6c734a3ee3bc0037f228c117af776f290c 100644
--- a/src/test/policy_block_tests.cpp
+++ b/src/test/policy_block_tests.cpp
@@ -45,6 +45,7 @@ static CBlock BlockWithMinerFund(const CChainParams &chainparams,
 BOOST_AUTO_TEST_CASE(policy_minerfund) {
     const CChainParams &chainparams = Params();
     const Consensus::Params &consensusParams = chainparams.GetConsensus();
+    Assert(m_node.args)->ForceSetArg("-enableminerfund", "1");
 
     std::array<BlockHash, 12> blockhashes;
     std::array<CBlockIndex, 12> blocks;
@@ -146,6 +147,8 @@ BOOST_AUTO_TEST_CASE(policy_minerfund) {
         checkMinerFundPolicy(BlockWithMinerFund(chainparams, amount),
                              cowperthwaiteBlockIndex, true);
     }
+
+    Assert(m_node.args)->ClearForcedArg("-enableminerfund");
 }
 
 BOOST_AUTO_TEST_SUITE_END()
